src/mpid/ch4/shm/src/shm_p2p.h

```diff
@@ -207,29 +299,103 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_SHM_mpi_irecv(void *buf, MPI_Aint count, MPI_
                                                  int rank, int tag, MPIR_Comm * comm,
                                                  int context_offset, MPIR_Request ** request)
 {
-    int ret;
-
+    int mpi_errno = MPI_SUCCESS;
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_SHM_MPI_IRECV);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_SHM_MPI_IRECV);
 
-    ret = MPIDI_POSIX_mpi_irecv(buf, count, datatype, rank, tag, comm, context_offset, request);
+#if (MPIDI_SHM_PT2PT_PROT == MPIDI_SHM_PT2PT_MULTIMODS)
+    MPIR_Comm *root_comm = NULL;
+    MPIR_Request *unexp_req = NULL;
+    MPIR_Context_id_t context_id = comm->recvcontext_id + context_offset;
+
+    /* When matches with an unexpected receive, it first tries to receive as
+     * a SHM optimized message (e.g., XPMEM SEND LMT). If fails, then receives
+     * as CH4 am message. Note that we maintain SHM optimized message in the
+     * same unexpected|posted queues as that used by CH4 am messages in order
+     * to ensure ordering.
+     */
+
+    /* Try to match with an unexpected receive request */
+    root_comm = MPIDIG_context_id_to_comm(context_id);
+    unexp_req = MPIDIG_dequeue_unexp(rank, tag, context_id, &MPIDIG_COMM(root_comm, unexp_list));
+
+    if (unexp_req) {
+        bool recvd_flag = false;
+
+        *request = unexp_req;
+        /* Mark as DEQUEUED so progress engine can complete a matched BUSY
+         * rreq once all data arrived */
+        MPIDIG_REQUEST(unexp_req, req->status) |= MPIDIG_REQ_UNEXP_DQUED;
+        MPIR_Comm_release(root_comm);   /* -1 for removing from unexp_list */
+
+        /* Try shmmod specific matched receive */
+        mpi_errno = MPIDI_SHM_mmods_try_matched_recv(buf, count, datatype, *request, &recvd_flag);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+
+        /* If not received, then fallback to POSIX matched receive */
+        if (!recvd_flag) {
+            mpi_errno = MPIDI_POSIX_mpi_imrecv(buf, count, datatype, *request);
+            if (mpi_errno)
+                MPIR_ERR_POP(mpi_errno);
+        }
+    } else {
+        /* No matching request found, post the receive request  */
+        MPIR_Request *rreq = NULL;
+
+        rreq = MPIDIG_request_create(MPIR_REQUEST_KIND__RECV, 2);
+        MPIR_ERR_CHKANDSTMT(rreq == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+
+        MPIR_Datatype_add_ref_if_not_builtin(datatype);
+        mpi_errno = MPIDIG_prepare_recv_req(rank, tag, context_id, buf, count, datatype, rreq);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+
+        MPIR_Comm_add_ref(root_comm);   /* +1 for queuing into posted_list */
+        MPIDIG_enqueue_posted(rreq, &MPIDIG_COMM(root_comm, posted_list));
+
+        *request = rreq;
```

Comment:
Please add MPIDI_POSIX_EAGER_RECV_POSTED_HOOK after posts a recv request.

Reply:
I am not sure why we should add `MPIDI_POSIX_EAGER_RECV_POSTED_HOOK` after calling `MPIDIG_enqueue_posted`. Could you please explain ?

Reply:
This branch of code is substitute of the call of `MPIDI_POSIX_mpi_irecv` function. But the `MPIDI_POSIX_mpi_irecv` does that notification after posts a recv request:

https://github.com/pmodels/mpich/blob/master/src/mpid/ch4/shm/posix/posix_recv.h#L67

The POSIX-EAGER transport must be notified about posted recv request because an advanced implementation of POSIX-EAGER transport might retards of handling of unexpected message while the corresponding recv request will not be posted.

Reply:
After reading the internal code of `MPIDI_POSIX_EAGER_RECV_POSTED_HOOK`, now my understanding is that the hook is used to hint the `posix/eager/fbox` transport that user is expecting an message from this rank, so the performance might be better if fbox polls from this rank at the next progress polling (i.e., in `MPIDI_POSIX_eager_recv_begin`). Is my understanding correct ?

Since we require the external code of `posix` to call this hook whenever post a recv request that will be handled by `posix`, we should have documentation for it. Moreover, it might be better to rename the interface to `MPIDI_POSIX_recv_posted_hook`, because the caller might not know it is an **eager** request or not. 

If you agree, I can create a PR to address the above changes and then add the call in this PR.

Reply:
@minsii Yes, you understanding is correct. Yes, renaming may make sense.

Reply:
Addressed this change in #3680 and pushed change in this PR (see last commit)